XLS-0045-prepublish-validator-lists/README.md

```diff
@@ -0,0 +1,824 @@
+
+##### Fields
+
+<Any explanatory text about the fields overall>
+
+---
+
+| Field Name        | Required?        |  JSON Type      | Internal Type     |
+|-------------------|:----------------:|:---------------:|:-----------------:|
+| `<field name>` | :heavy_check_mark: | `<string, number, object, array, boolean>` | `<UINT128, UINT160, UINT256, ...>` |
+
+<Any explanatory text about specific fields>
+
+###### Flags
+
+> | Flag Name            | Flag Value  | Description |
+>|:---------------------:|:-----------:|:------------|
+>| `lsf<flag name>` | `0x0001`| <flag description> |
+
+<Any explanatory text about specific flags>
+-->
+
+### Overview of changes
+
+1. Add functionality to the UNL to include a future effective date. Change
+the published file format to allow multiple UNLs to be returned.
+(Limit number of future UNLs to mitigate some potential attacks.) All
+nodes will start using the new UNL after the first validated ledger with
+an earlier close time (falling back to the wall clock if necessary, such
+as when the node is not synced).
+2. Add enforcement of the expiration date by changing rippled's behavior
+to something similar to being amendment blocked (reuse that code as much
+as possible).
+
+### Data format
+
+#### Glossary
+
+Term or acronym | Definition
+----------------|------------
+VL  | **V**alidator **L**ist. Also known as the "UNL", this is the list of validators that a server trusts not to collude.
+UNL | **U**nique **N**ode **L**ist. Also known as the "VL", this is the list of validators that a server trusts not to collude.
+UNL file (or VL file) | A file containing a publisher's manifest, public key, etc., and one or more UNLs.
+
+#### Current format
+
+The current format of the VL file (e.g. as downloaded from
+https://vl.ripple.com/) is
+
+```
+{
+  "public_key" : string representing the publisher's hex-encoded master
+    public key,
+  "manifest" : string representing the base-64 or hex-encoded manifest
+    containing the publisher's master and signing public keys,
+  "version": unsigned integer representing the VL format version. Currently
+    only 1 is valid,
+  "signature": string representing the hex-encoded signature of the blob
+    using the publisher's signing key,
+  "blob": string representing base-64 encoded JSON containing the actual list.
+    See below
+}
+```
+
+The `blob` format is
+
+```
+{
+  "sequence" : Unsigned integer sequence of this VL. The sequence number
+    must increase monotonically. More specifically, validator lists with
+    sequences less than the current are ignored. The special sequence
+    UInt(-1) indicates that the master key is revoked,
+  "expiration" : Unsigned integer representing the ripple time point when
+    this list will no longer be valid,
+  "validators" : [
+    {
+      "validation_public_key" : hex-encoded master public key,
+      "manifest" : (optional) base64 or hex-encoded validator manifest
+    },
+    ...Optionally repeats...
+  ]
+}
+```
+
```

Comment:
```suggestion

"[ripple time](https://xrpl.org/docs/references/protocol/data-types/basic-data-types/#specifying-time)"
is the number of seconds since the "Ripple Epoch" of January 1, 2000 (00:00
UTC). This is similar to how the [Unix epoch](https://en.wikipedia.org/wiki/Unix_time)
works, except the Ripple Epoch is 946,684,800 seconds after the Unix Epoch. Do
not convert Ripple Epoch times to Unix Epoch times in 32-bit variables, as this
could lead to integer overflows.

```